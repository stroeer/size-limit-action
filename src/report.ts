import fs from 'fs';
import { resolve } from 'path';
import * as artifact from '@actions/artifact';
import { mkdirP } from '@actions/io';
import { getPaths, getConfigs, getSizes } from './createReport';
import diffSizes from './diffSizes';
import { Sizes, Changes } from './types';

const ACTION_TMP_DIR = resolve(__dirname, '../.size-limit-action-temp-files');
const UPLOADS_DIR = resolve(ACTION_TMP_DIR, 'uploads');
const DOWNLOADS_DIR = resolve(ACTION_TMP_DIR, 'downloads');

export async function save(name: string, sizes: Sizes): Promise<void> {
  const client = artifact.create();
  const filePath = resolve(UPLOADS_DIR, `${name}.json`);
  await mkdirP(UPLOADS_DIR);
  await fs.promises.writeFile(filePath, JSON.stringify(sizes), 'utf8');
  await client.uploadArtifact(name, [filePath], UPLOADS_DIR, { continueOnError: false });
}

// todo proper error handling, what if the import() fails
// todo the currently logged warning in the catch(error) should be part of the comment footer
export async function load(name: string): Promise<Sizes> {
  const client = artifact.create();
  try {
    await client.downloadArtifact(name, DOWNLOADS_DIR);
    const module = await import(resolve(UPLOADS_DIR, `${name}.json`));
    return module.default;
  } catch (error) {
    console.warn(
      `Warning: Failed to load saved sizes for "${name}" (${error.message}). Falling back an empty set of sizes, meaning all diffs are a 100% increase.`,
    );
    return {};
  }
}

export async function create(pattern: string, cwd?: string): Promise<Sizes> {
  const paths = await getPaths(pattern, cwd);
  const configs = await getConfigs(paths);
  return await getSizes(configs);
}

export async function diff(oldSizes: Sizes, newSizes: Sizes): Promise<Changes> {
  const paths = Object.keys(newSizes);
  const changes = paths.map(path => diffSizes(path, oldSizes[path], newSizes[path]));
  return Object.fromEntries(changes);
}
