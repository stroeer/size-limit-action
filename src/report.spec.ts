import { diff } from './report';

describe('report', () => {
  const file5 = { path: 'file5.js', bytes: 5 };
  const file10 = { path: 'file10.js', bytes: 10 };

  const incFile5 = { ...file5, bytes: 10 };
  const incFile10 = { ...file10, bytes: 20 };

  const decFile5 = { ...file5, bytes: 1 };
  const decFile10 = { ...file10, bytes: 5 };

  test.each`
    name                                                             | oldSizes                      | newSizes
    ${'an empty object for empty inputs'}                            | ${{}}                         | ${{}}
    ${'files as extra bytes'}                                        | ${{}}                         | ${{ myPkg: [file5, file10] }}
    ${'existing files as unchanged'}                                 | ${{ myPkg: [file5, file10] }} | ${{ myPkg: [file5, file10] }}
    ${'removed files missing bytes'}                                 | ${{ myPkg: [file5, file10] }} | ${{ myPkg: [] }}
    ${'additions in bytes and percentages when sizes have changed'}  | ${{ myPkg: [file5, file10] }} | ${{ myPkg: [incFile5, incFile10] }}
    ${'reductions in bytes and percentages when sizes have changed'} | ${{ myPkg: [file5, file10] }} | ${{ myPkg: [decFile5, decFile10] }}
    ${'reductions and additions at the same time'}                   | ${{ myPkg: [file5, file10] }} | ${{ myPkg: [decFile5, incFile10] }}
  `('diff returns $name', async ({ oldSizes, newSizes }) => {
    await expect(diff(oldSizes, newSizes)).resolves.toMatchSnapshot();
  });
});
