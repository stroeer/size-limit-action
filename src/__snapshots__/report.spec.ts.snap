// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`report diff returns additions in bytes and percentages when sizes have changed 1`] = `
Object {
  "myPkg": Object {
    "files": Array [
      Object {
        "bytes": 10,
        "change": 5,
        "path": "file5.js",
        "percent": 100,
      },
      Object {
        "bytes": 20,
        "change": 10,
        "path": "file10.js",
        "percent": 100,
      },
    ],
    "total": Object {
      "bytes": 30,
      "change": 15,
      "path": "myPkg",
      "percent": 100,
    },
  },
}
`;

exports[`report diff returns an empty object for empty inputs 1`] = `Object {}`;

exports[`report diff returns existing files as unchanged 1`] = `
Object {
  "myPkg": Object {
    "files": Array [
      Object {
        "bytes": 5,
        "change": 0,
        "path": "file5.js",
        "percent": 0,
      },
      Object {
        "bytes": 10,
        "change": 0,
        "path": "file10.js",
        "percent": 0,
      },
    ],
    "total": Object {
      "bytes": 15,
      "change": 0,
      "path": "myPkg",
      "percent": 0,
    },
  },
}
`;

exports[`report diff returns files as extra bytes 1`] = `
Object {
  "myPkg": Object {
    "files": Array [
      Object {
        "bytes": 5,
        "change": 5,
        "path": "file5.js",
        "percent": 100,
      },
      Object {
        "bytes": 10,
        "change": 10,
        "path": "file10.js",
        "percent": 100,
      },
    ],
    "total": Object {
      "bytes": 15,
      "change": 15,
      "path": "myPkg",
      "percent": 100,
    },
  },
}
`;

exports[`report diff returns reductions and additions at the same time 1`] = `
Object {
  "myPkg": Object {
    "files": Array [
      Object {
        "bytes": 1,
        "change": -4,
        "path": "file5.js",
        "percent": -80,
      },
      Object {
        "bytes": 20,
        "change": 10,
        "path": "file10.js",
        "percent": 100,
      },
    ],
    "total": Object {
      "bytes": 21,
      "change": 6,
      "path": "myPkg",
      "percent": 40,
    },
  },
}
`;

exports[`report diff returns reductions in bytes and percentages when sizes have changed 1`] = `
Object {
  "myPkg": Object {
    "files": Array [
      Object {
        "bytes": 1,
        "change": -4,
        "path": "file5.js",
        "percent": -80,
      },
      Object {
        "bytes": 5,
        "change": -5,
        "path": "file10.js",
        "percent": -50,
      },
    ],
    "total": Object {
      "bytes": 6,
      "change": -9,
      "path": "myPkg",
      "percent": -60,
    },
  },
}
`;

exports[`report diff returns removed files missing bytes 1`] = `
Object {
  "myPkg": Object {
    "files": Array [
      Object {
        "bytes": 0,
        "change": -5,
        "path": "file5.js",
        "percent": -100,
      },
      Object {
        "bytes": 0,
        "change": -10,
        "path": "file10.js",
        "percent": -100,
      },
    ],
    "total": Object {
      "bytes": 0,
      "change": -15,
      "path": "myPkg",
      "percent": -100,
    },
  },
}
`;
